ogs6py/log_parser/common_ogs_analyses.py

```diff
@@ -0,0 +1,176 @@
+import pandas as pd
+import numpy as np
+
+
+# Helper functions
+def check_input(df, interest, context):
+    diff = set(interest) - set(df.columns)
+    if diff:
+        raise Exception('Column(s) of interest ({}) is/are not present in table'.format(','.join(diff)))
+    diff = set(context) - set(df.columns)
+    if diff:
+        raise Exception('Column(s) of context ({}) is/are not present in table', ','.format(','.join(diff)))
+
+
+def check_output(pt, interest, context):
+    if pt.empty:
+        raise Exception(
+            'The values of {} are not associated to all of {}. Call or see fill_ogs_context'.format(','.join(interest),
+                                                                                                    ','.join(context)))
+
+
+# decorator for analyses
+def pre_post_check(interest, context):
+    def wrap(f):
+        def wrapped_f(df):
+            check_input(df, interest, context)
+            pt = f(df)
+            check_output(pt, interest, context)
+            return pt
+
+        return wrapped_f
+
+    return wrap
+
+
+'''
+Analysis with focus on computation time per time step. It combines time step specific measurements 'output time'
+and 'time step solution time' with iteration specific measurements 'assembly time', 'linear solver time', 'Dirichlet time'.
+Time from iteration are accumulated.
+'''
+
+
+def analysis_time_step(df):
+    interest1 = ['output_time', 'time_step_solution_time']
+    interest2 = ['assembly_time', 'linear_solver_time', 'dirichlet_time']
+    interest = [*interest1, *interest2]
+    context = ['mpi_process', 'time_step']
+    check_input(df, interest, context)
+
+    dfe_ts = df.pivot_table(interest1, context)
+    # accumulates coupling iterations and newton iterations
+    dfe_tsi = df.pivot_table(interest2, context, aggfunc='sum')
+
+    dfe = dfe_ts.merge(dfe_tsi, left_index=True, right_index=True)
+    check_output(dfe, interest, context)
+    return dfe
+
+
+def analysis_simulation(df):
+    interest = ['execution_time']  # 'start_time'
+    context = ['mpi_process']
+    check_input(df, interest, context)
+
+    pt = df.pivot_table(interest, context)
+    check_output(pt, interest, context)
+    return pt
+
+
+def analysis_convergence_newton_iteration(df):
+    dfe_newton_iteration = df.copy()
+    interest = ['dx', 'x', 'dx_x']
+    if 'coupling_iteration' in df:
+        context = ['time_step', 'coupling_iteration', 'process',
+                   'iteration_number', 'component']
```

Comment:
I think the 'component' keyword optional, it is only there if a multicomponent variable like displacement is involved.

Reply:
You are right, that the _component_ is uninteresting when you have scalar processes or at least scalar convergence criteria, but if it is not so, then it is helpful to have it. IMHO having one variable more does not justify to create a flag or another analysis, so I would not change this.

Reply:
Agree with Dominik here in terms that we do not add another analysis. I would treat the functions in common_ogs_analysis as examples and good starting point for specific/custom analyses. In this special case here the study without component seems to be more generic and would follow Joergs proposal to remove component.

Reply:
We could check whether `'component'` is there in the data and add it conditionally to the context.